pkgs/development/compilers/rust/binary.nix

```
@@ -46,6 +46,12 @@ rec {
         patchelf \
           --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
           "$out/bin/rustc"
+        ${optionalString (stdenv.lib.versionAtLeast version "1.46")
+        # rustc bootstrap needs libz starting from 1.46
+        ''
+          ln -s ${zlib}/lib/libz.so.1 $out/lib/libz.so.1
+          ln -s ${zlib}/lib/libz.so $out/lib/libz.so
+        ''}
         patchelf \
           --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
           "$out/bin/rustdoc"
```

What you have now still introduces a blank line, however, I would do
```suggestion
        patchelf \
          --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
          "$out/bin/rustc"
      '' + optionalString (stdenv.lib.versionAtLeast version "1.46") 
        # rustc bootstrap needs libz starting from 1.46
        ''
          ln -s ${zlib}/lib/libz.so.1 $out/lib/libz.so.1
          ln -s ${zlib}/lib/libz.so $out/lib/libz.so
        ''
      + ''
        patchelf \
          --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
          "$out/bin/rustdoc"
```

and then parens around the all concatenated thing (which I cannot include in the suggestion, sadly).